src/google/cloud/ndb/tasklets.py

```diff
@@ -248,14 +247,17 @@ def __init__(self, generator, context, info="Unknown"):
     def _advance_tasklet(self, send_value=None, error=None):
         """Advance a tasklet one step by sending in a value or error."""
         try:
-            with self.context:
+            with self.context.use():
                 # Send the next value or exception into the generator
                 if error:
                     self.generator.throw(type(error), error)
 
                 # send_value will be None if this is the first time
                 yielded = self.generator.send(send_value)
 
+                # Context may have changed in tasklet
```

Comment:
This is the one liner change that actually handles the possibility that a context might change internally to a tasklet. All the other refactoring is done to support this one-liner change.